File: 071_simplify_path/Solution.go
package main

import (
    "fmt"
    "path/filepath"
    "strings"
)

func simplifyPath(path string) string {
    return filepath.Clean(path)
}

func simplifyPath2(path string) string {
    pathArr := strings.Split(path, "/")
    var stack []string
    for _, tmp := range pathArr {
        if tmp == "" || tmp == "." { continue }
        if tmp == ".." {
            if len(stack) != 0 {
                stack = stack[:len(stack) - 1]
            }
            continue
        }
        stack = append(stack, tmp)
    }

    return "/" + strings.Join(stack, "/")
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

func main() {
    testCases := []string {
        "/...",
        "/home/",
        "/../",
        "/home//foo/",
        "/a/../../b/../c//.//",
        "/a//b////c/d//././/..",
        "/a/./b/../../c/",
        "/a//b////c/d//././/..",
    }

    for _, test := range testCases {
        fmt.Println(simplifyPath2(test))
    }
}

File: 071_simplify_path/Solution.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        path_list = path.split('/')
        result = []
        for tmp in path_list:
            if tmp not in ['', '.', '..']:
                result.append(tmp)
            elif tmp == '..' and result:
                result.pop()

        return '/' + '/'.join(result)


if __name__ == '__main__':
    tests = [
        "/home/",
        "/../",
        "/home//foo/",
        "/a/../../b/../c//.//",
        "/a//b////c/d//././/.."
    ]

    solution = Solution()
    for test in tests:
        print(solution.simplifyPath(test))

